pageserver/page_api/src/client.rs

```diff
@@ -0,0 +1,234 @@
+    auth_header: Option<String>,
+}
+
+fn convert_metadata_err(e: InvalidMetadataValue, s: String) -> Error {
+    let io_err = Error::new(
+        ErrorKind::InvalidInput,
+        format!("{} header was invalid: {}", s, e),
+    );
+    io_err
+}
+
+impl Client {
+
+    pub async fn new(connstring: String,
+                    tenant_id: TenantId,
+                    timeline_id: TimelineId,
+                    shard_id: ShardIndex,
+                    auth_header: Option<String>) -> Result<Self, Error> {
+
+        let mut client = proto::PageServiceClient::connect(connstring).await
+            .map_err(|e| {
+                Error::new(ErrorKind::Other, format!("Failed to connect to page service: {}", e))
+            })?;
+
+        // Attempt to parse headers, so that parsing fails early. Later, we can assume
+        // that the headers are valid ASCII strings.
+        let _tenant_ascii : AsciiMetadataValue = tenant_id.to_string().try_into()
+            .map_err(|e: InvalidMetadataValue| {
+                convert_metadata_err(InvalidMetadataValue::from(e), "tenant-id".to_string())
+            })?;
+
+        let _timeline_ascii : AsciiMetadataValue = timeline_id.to_string().try_into()
+            .map_err(|e: InvalidMetadataValue| {
+                convert_metadata_err(InvalidMetadataValue::from(e), "timeline-id".to_string())
+            })?;
+
+        let _shard_ascii : AsciiMetadataValue = shard_id.clone().to_string().try_into()
+            .map_err(|e: InvalidMetadataValue| {
+                convert_metadata_err(InvalidMetadataValue::from(e), "shard-id".to_string())
+            })?;
+
+        Ok(Self {
+            client,
+            tenant_id: tenant_id.to_string(),
+            timeline_id: timeline_id.to_string(),
+            shard_id: shard_id.to_string(),
+            auth_header,
+        })
+    }
+
+    // helper to insert metadata headers
+    fn insert_metadata<T>(self, mut req: tonic::Request<T>) -> (tonic::Request<T>, Self) {
+        let metadata = req.metadata_mut();
+
+        // Assume that the headers are valid Ascii strings, since we already checked them in the constructor.
+        let tenant_ascii : AsciiMetadataValue = self.tenant_id.clone().try_into().unwrap();
+        let timeline_ascii : AsciiMetadataValue = self.timeline_id.clone().try_into().unwrap();
+        let shard_ascii : AsciiMetadataValue = self.shard_id.clone().try_into().unwrap();
+
+        metadata.insert("neon-tenant-id", tenant_ascii);
+        metadata.insert("neon-timeline-id", timeline_ascii);
+        metadata.insert("neon-shard-id", shard_ascii);
+
+        return (req, self);
+    }
+
+    pub async fn check_rel_exists(
+        mut self,
+        req: model::CheckRelExistsRequest,
+    ) -> Result<model::CheckRelExistsResponse, tonic::Status> {
+
+        let proto_req = proto::CheckRelExistsRequest {
+            read_lsn: Some(req.read_lsn.try_into()?),
+            rel: Some(req.rel.into()),
+        };
```

Comment:
The domain types already have built-in conversions for these: `let req: proto::CheckRelExistsRequest = req.try_into()?`.